pkg/workflow/providers/oam/apply.go

```diff
@@ -117,6 +129,54 @@ func (p *provider) ApplyComponent(ctx wfContext.Context, v *value.Value, act wfT
 	return nil
 }
 
+// ApplyComponent apply component.
+func (p *provider) ApplyComponent(ctx wfContext.Context, v *value.Value, act wfTypes.Action) error {
+	return p.applyComponent(ctx, v, act, nil)
+}
+
+// ApplyComponents apply components in parallel.
+func (p *provider) ApplyComponents(ctx wfContext.Context, v *value.Value, act wfTypes.Action) error {
+	components, err := v.LookupValue("components")
+	if err != nil {
+		return err
+	}
+	parallelism, err := v.GetInt64("parallelism")
+	if err != nil {
+		return err
+	}
+	if parallelism <= 0 {
+		return errors.Errorf("parallelism cannot be smaller than 1")
+	}
+	mu := &sync.Mutex{}
+	var wg sync.WaitGroup
+	ch := make(chan struct{}, parallelism)
+	var errs errors2.ErrorList
+	err = components.StepByFields(func(name string, in *value.Value) (bool, error) {
+		ch <- struct{}{}
+		wg.Add(1)
+		go func(_name string, _in *value.Value) {
+			defer func() {
+				wg.Done()
+				<-ch
+			}()
+			if err := p.applyComponent(ctx, _in, act, mu); err != nil {
+				mu.Lock()
+				errs = append(errs, errors.Wrapf(err, "failed to apply component %s", _name))
+				mu.Unlock()
```

Comment:
you can using a channel to send signal for error or success, then you can avoid using waitGroup and this mutex. Just receive the signal outside.

Reply:
This mutex is not only used in L163 and L165. It is also used inside applyComponent while handling ResourceTracker. I think channel can handle the return of error but not be able to handle the concurrency problem inside applyComponent.

Reply:
at lease we don't need the channel here now.